org-roam.el

```diff
@@ -617,28 +618,39 @@ Throw an error if multiple choices exist."
      (t
       (user-error "Multiple nodes exist with title or alias \"%s\"" s)))))
 
+(defvar org-roam--db-prepped nil)
+
+(defun org-roam--sql (connection sql-string)
+  "Query emacsql-managed CONNECTION with low-level SQL-STRING"
+  (emacsql-clear connection)
+  (emacsql-send-message connection sql-string)
+  (emacsql-wait connection)
+  (emacsql-parse connection))
+
+(defun org-roam--db-prep ()
+  (if (not org-roam--db-prepped)
+      (progn
+        (setq org-roam--db-prepped t)
+        (org-roam--sql (org-roam-db) "
+    create index if not exists nodes_id on nodes (id);
+    create index if not exists aliases_id on aliases (node_id);"))))
+
+;; THIS MUST BE IN SYNC WITH ORG-ROAM-NODE DEF!!!
+(defun org-roam-query-nodes ()
+  "Get nodes from database"
+  (org-roam--db-prep)
+  (mapcar #'car (org-roam--sql (org-roam-db) "select '#s(org-roam-node ' || id || ' ' || file || ' nil ' || pos || ' nil nil nil nil ' || title || ' ' || properties || ' ' || ifnull(olp, 'nil') || ' nil nil nil' || ')' from nodes union select '#s(org-roam-node ' || node_id || ' ' || nodes.file || ' nil ' || pos || ' nil nil nil nil ' || alias || ' ' || properties || ' ' || ifnull(olp, 'nil') || ' nil nil nil' || ')' from aliases left join nodes on node_id = id;")))
```

Comment:
Can we profile how much perf improvements come from manually constructing the query like this?

If the bulk of the perf improvements come from the indexes on the alias tables for the joins, then I'm leaning towards keeping query as is to keep the code clean.

The caching of the display format looks much more acceptable.

Reply:
Sure, I can check what difference it makes

Reply:
It's a good thing you asked that. You are correct that the index is what makes the difference. In fact, besides being harder to read and maintain, the inlined SQL version is actually around 7% slower than the original! I'm guessing that SQLite's string concatenation isn't as well implemented as Emacs'.